File: 03.longest-substring-without-repeating-characters/LongestSubstring.go
package main

import "fmt"

func longestSubstring(s string) (string, int) {

	locations := make([]int, 256)
	for i := range locations {
		locations[i] = -1
	}

	index, maxLength, left := 0, 0, 0
	for i, value := range s {
		if locations[value] >= left {
			left = locations[value] + 1
		} else if i-left+1 > maxLength {
			index = i
			maxLength = index - left + 1
		}
		locations[value] = i
	}

	return s[index-maxLength+1 : index+1], maxLength
}

func solveProblem(s string) {
	substr, length := longestSubstring(s)
	fmt.Printf("string: (%s), substr: (%s), length=%d\n", s, substr, length)
}

func main() {
	solveProblem("abcabcbb")
	solveProblem("bbbbb")
	solveProblem("pwwkew")
}

File: 03.longest-substring-without-repeating-characters/LongestSubstring.py

def calc_longest_substr(s : str) -> tuple[str, int]:
    locations = [-1 for x in range(255)]

    index, max_length, left = 0, 0, 0
    for i, c in enumerate(s):       
        if locations[ord(c)] >= left:
            left = locations[ord(c)] + 1
        elif i - left + 1 > max_length:
            max_length = i - left + 1
            index = i
        locations[ord(c)] = i
    
    return s[index - max_length + 1 : index + 1], max_length


def solve_problem(s:str) -> None:
    substr, length = calc_longest_substr(s)
    print(f'string: {s}, longest substr: {substr}, length: {length}')


if __name__ == '__main__':
    solve_problem("abcabcbb")
    solve_problem("bbbbb")
    solve_problem("pwwkew")

